### ISSUE TITLE
[1.1.2+] Models' key type forced to string when using System\Models\File attachment in Models

### ISSUE THREAD

**Issue by adrenth** [octobercms/october#5541]
- OctoberCMS Build: 1.1.2+
- Database Engine: MySQL
- Plugins Installed: `Vdlp.Acme` (https://github.com/vdlp/octobercms-issues/tree/octobercms-library-PR502)

### Description:

Since the merge of PR https://github.com/octobercms/library/pull/502 we got a lot of type errors in our applications which uses the `AttachOne` and `AttachMany` relationships. This occurs only when having a relation to to the model `System\Models\File`.

In the following commit a property keyType was added to the `$morphTo` relation definition with a default value `string` (see: https://github.com/octobercms/library/pull/502/files#diff-76ff8fa0729770c910358df5076f363c4214d00da4c12c41195525123ba1aec3R31-R33)

The `keyType` property will be read in the `AttachOne` and `AttachMany` classes and use this new property to set the `keyType` on the **parent** model.

By calling `setKeyType` on the parent model the key type will be invalidly set to `string`. This causes the integer ID to be cast as `string` and breaking code down the line. In following case the `keyType` of our `Acme` model is `int`. After loading the `attachment` relation, our `keyType` is switched to string and `$acme->id` will return `(string) "1"`. 

We always develop in strict mode and this causes issues when having code like this:

```php
public function getAcmeModelById(int $id): Acme
{
    return Acme::query()->find($id);
}

// This works
$acmeOne = Acme::query()->findOrFail(1);
getAcmeModelById($acmeOne->id);

// This will throw a Type error exception!
$acmeTwo = Acme::query()->findOrFail(1);
$acmeTwo->attachment;

getAcmeModelById($acmeTwo->id); 
```

Although this code change might fix the issue for PostgreSQL users, but this is a breaking change. This will cause type errors in code which rely on the Models' key type. And also when change the `keyType` of `$morphTo` to `int` won’t change anything at all. The key will always be returned as a `string`.

**Our proposal would be to revert this [pull request](https://github.com/octobercms/library/pull/502). And let's look into this PostgreSQL issue further (issue reporter @RomainMazB).**

### Steps To Reproduce:

Use October CMS version 1.1.2 (or 1.1.3).

- Download the `Vdlp.Acme` plugin from https://github.com/vdlp/octobercms-issues/tree/octobercms-library-PR502. 
- Put it in the plugin directory `plugins/vdlp/acme`.
- Run command `php artisan october:up`
- Then run the command `php artisan vdlp:acme:test` to run the example code.

Example code:

```php
<?php

declare(strict_types=1);

namespace Vdlp\Acme\Console;

use Illuminate\Console\Command;
use System\Models\File;
use Vdlp\Acme\Models\Acme;

final class TestCommand extends Command
{
    public function __construct()
    {
        $this->signature = 'vdlp:acme:test';
        $this->description = 'A test command';

        parent::__construct();
    }

    public function handle(): void
    {
        $identifier = $this->createExample();

        $this->loadExample($identifier);
    }

    private function createExample(): int
    {
        // Create a new Acme model

        $acme = Acme::create([
            'name' => 'my acme',
        ]);

        // Create a new File model

        $file = new File();
        $file->fromUrl('https://www.vdlp.nl/storage/app/media/technologie/letsmanage10.jpg', 'somefilename.jpg');

        // Get the key

        var_dump($acme->getKey()); // This will output: int(1)

        // Attach the File model to Acme model

        $acme->attachment()->add($file);

        // Get the key

        var_dump($acme->getKey()); // This will output: string(1) "1"

        return (int) $acme->getKey();
    }

    private function loadExample(int $identifier): void
    {
        $newAcme = Acme::query()->findOrFail($identifier);

        var_dump($newAcme->getKeyType()); // string(3) "int"

        $newAcme->attachment;

        var_dump($newAcme->getKeyType()); // string(6) "string"
    }
}
```

**Comment by daftspunk** [comment]
Thanks for the detailed report @adrenth. This is a concerning addition as it addresses the symptom and not the underlying cause. It may have also broken MySQL's internal performance handling. We will roll it back and ship 1.1.4 today

As a side note, the next major release of October CMS has undergone a complete code audit and this change was not carried across either. We have a more robust solution planned to address this

**Comment by daftspunk** [comment]
Test completed OK

```
λ php artisan vdlp:acme:test
int(1)
int(1)
string(3) "int"
string(3) "int"
```

**Comment by daftspunk** [comment]
This will be fixed in patch release 1.1.4

Thanks for using October CMS :pray:
